feat: Update Image block failed upload visuals

derekblank opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Related PRs

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6491
* https://github.com/wordpress-mobile/WordPress-iOS/pull/22282
* https://github.com/wordpress-mobile/WordPress-Android/pull/19878

## What?
Display a custom message for failed media uploads that are paused due to the lack of an internet connection. 

## Why?
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/6407. Improve the UX by better communicating to users why the media upload is failed so that they may take action. 

## How?
Integrate a new "paused" media upload state that is sent by the host app. For block using `enablePausedUploads` (currently, only Image blocks), a specific "Waiting for connection" error message is displayed. 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Add an Image, Gallery, or Media + Text block. 
2. Disable your devices network connection (e.g. airplane mode). 
3. Attempt to attach image(s) to the block. 
4. Verify the "Waiting for connection" error message is displayed. 
5. Repeat steps 1–4 changing when the network connection is disabled, e.g., midway through an upload. 

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
Attempt to complete the aforementioned testing instructions via a mobile device's screen reader. 

## Screenshots or screencast <!-- if applicable -->

<details><summary>iOS</summary>

| Light mode | Dark mode |
| - | - |
| ![iOS image block failed upload design in light mode](https://github.com/WordPress/gutenberg/assets/438664/7c5c407e-9b16-4e97-8333-54bbfa774602) | ![iOS image block failed upload design in dark mode](https://github.com/WordPress/gutenberg/assets/438664/0c049af8-9316-4871-bd8c-ec90ff49181b) |

</details>

<details><summary>Android</summary>

| Light mode | Dark mode |
| - | - |
| ![Android image block failed upload design in light mode](https://github.com/WordPress/gutenberg/assets/438664/3e2e8b9f-a542-49df-bfba-31a1ea2afacd) | ![Android image block failed upload design in dark mode](https://github.com/WordPress/gutenberg/assets/438664/fb33247d-68fd-485c-9a91-7cb8c319f05c) |

</details>

derekblank committed:
Add subscribeConnectionStatus to MediaUploadProgress component

dcalhoun commented on packages/react-native-bridge/index.js:
```diff
@@ -226,7 +226,7 @@ export function useIsConnected() {
 }
 
-function subscribeConnectionStatus( callback ) {
+export function subscribeConnectionStatus( callback ) {
```
An alternative to exporting this utility would be providing an higher-order component for class components. This practice is [fairly prevalent](https://github.com/WordPress/gutenberg/blob/93349ff2ba3010baf4957f575a9a6a8672ec98c5/packages/data/src/components/with-select/index.js) in the Gutenberg project today. 

There may be value in aligning with project practices and avoiding duplicative logic. WDYT? 

```js
/**
 * WordPress dependencies
 */
import { createHigherOrderComponent } from "@wordpress/compose";

/**
 * Internal dependencies
 */
import { useIsConnected } from "./index";

export default createHigherOrderComponent( ( WrappedComponent ) => {
	return ( props ) => {
		const isConnected = useIsConnected();
		return <WrappedComponent { ...props } isConnected={ isConnected } />;
	};
}, 'withIsConnected' );
```

```js
function MyClassComponent( props ) {
	return props.isConnected ? (
		<Text>Connected</Text>
	) : (
		<Text>Not Connected</Text>
	);
}

export default withIsConnected( MyClassComponent );
```

derekblank replied:
I definitely went down the rabbit hole for using a HOC here (and using Context, too). I am not against the HOC at all, but would welcome your feedback on the thought that kept me from implementing a HOC: do you perceive that we (as in, anyone that works on Gutenberg, and particularly Gutenberg Mobile) have a general mission to update class components to function components? I have the understanding that class components would eventually need to be refactored, and that implementing any further HOCs may discourage us from doing so. I am not certain if this is my own bias as I vastly prefer to work with function components and hooks. I do recognize that a HOC would allow us to benefit from code reuse in this case. 👍

dcalhoun replied:
I do not perceive a HOC created for the purposes of sharing logic abstracted into a Hook to be a negative thing or a "crutch" discouraging migration. I perceive it to be a utility to help the team retain velocity against the highest priorities, whatever that may be at a given time. 

I do believe the project intends to migrate class components to function components. I perceive it is a judgement call as to the appropriate time do so. Generally, I recommend migrating components first in a separate PR before modifying them. 

All that to say, we should use a HOC here. It allows us to reuse logic that we would otherwise have to repeat and ensure we keep in sync to avoid bugs.

derekblank replied:
Added `withIsConnected` as a HOC, and opened a PR in https://github.com/WordPress/gutenberg/pull/56966 so that it could be merged independently from this work. I've updated this PR to `withIsConnected`, however. Will note a few other things in the comments.

dcalhoun commented on packages/block-editor/src/components/media-upload-progress/index.native.js:
```diff
@@ -36,10 +39,18 @@ export class MediaUploadProgress extends Component {
 	componentDidMount() {
 		this.addMediaUploadListener();
+
+		this.subscription = subscribeConnectionStatus(
+			this.handleConnectionStatusChange
+		);
```
Noting that we if we retain the approach of _not_ using the `isConnected` Hook, we'll need to update this `componentDidMount` to fetch the initial state of the network connection, similar to the Hook.

https://github.com/WordPress/gutenberg/blob/aadf8c66ff345b176041ad82e3793191ade5d271/packages/react-native-bridge/index.js#L196-L212

derekblank committed:
Update MediaUploadProgress to use withIsConnected HOC

derekblank commented on packages/block-editor/src/components/media-upload-progress/index.native.js:
```diff
@@ -115,19 +117,26 @@ export class MediaUploadProgress extends Component {
+			// eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace
+			return __( 'Failed to insert media.\nTap for more info.' );
+		}
+		return __( 'You are currently offline.' );
```
We may want to expand this message to include another line like "Media uploads will resume when network connectivity has been restored." Leaving it as "You are currently offline." until that feature is implemented. Open to other suggestions, too.

dcalhoun replied:
Yes, we might consider communicating that uploads will restart once network connectivity is re-established. 

In the meantime, the current message is duplicative of the top-level "Working offline" message provided by https://github.com/WordPress/gutenberg/pull/56934. We might consider using a more specific message. 
```suggestion
		return __( 'Upload paused while offline' );
```

Additionally, should we add the Wi-Fi icon referenced in the mocks?

derekblank commented on packages/block-editor/src/components/media-upload-progress/index.native.js:
```diff
@@ -115,19 +117,26 @@ export class MediaUploadProgress extends Component {
-			showSpinner || styles.progressBarHidden,
+			showSpinner && this.props.isConnected
+				? styles.progressBarHidden
+				: null,
```
In the current `trunk`, the entire progress bar becomes transparent via `styles.progressBarHidden`. When offline, this logic will persist the progressBar _track_, but as far as I can tell, the actual `progress` value is reset to `0` when `MEDIA_UPLOAD_STATE_FAILED` is thrown.

I have not yet uncovered where this progress reset occurs, but I believe it may be either in the host apps code or in the react-native-bridge glue code. Ideally we would not reset the progress value to `0` when `MEDIA_UPLOAD_STATE_FAILED` is thrown, but persist it and assume that the upload may be resumed. I also entertained the idea of introducing a new state like `MEDIA_UPLOAD_STATE_PAUSED` for when the app is offline, but that may be out of scope here.

dcalhoun replied:
While researching for https://github.com/wordpress-mobile/gutenberg-mobile/issues/6406, I encountered the potentially [relevant iOS code](https://github.com/wordpress-mobile/WordPress-iOS/blob/d6ec2c98cc52c591232935114cb94b9d8eccc0b7/WordPress/Classes/ViewRelated/Gutenberg/GutenbergMediaInserterHelper.swift#L244-L249). I have not yet research the Android side of the logic, but hopefully it is somewhat similar. 

From merely glancing at that logic, it would appear it already conditionally resets the progress state based on the error type. I wonder if we might be able to modify it for our needs.

dcalhoun reviewed (commented):
FYI I rebased this branch atop `trunk` now that https://github.com/WordPress/gutenberg/pull/56966 merged.

dcalhoun committed:
refactor: Rename `useIsConnected` to `useNetworkConnectivity`

Attempt to better communicate the Hook intent.

dcalhoun committed:
test: Add `useNetworkConnectivity` Hook tests

dcalhoun committed:
docs: Document `useNetworkConnectivity` Hook

dcalhoun committed:
refactor: Rename `withIsConnected` to `withNetworkConnectivity`

dcalhoun committed:
fix: Optimistically presume network connectivity

Prior to making the asynchronous request to the host app across the
bridge, it is a better UX to presume network connectivity is present
rather than displaying network connectivity messages briefly.

dcalhoun committed:
test: Create realistic default `requestConnectionStatus` mock

dcalhoun committed:
fix: Prevent duplicative offline status indicators

Hoist the `OfflineStatus` indicator from the block list to the editor.
The block list is leveraged for inner blocks, which means it rendered
nested `OfflineStatus` indicators for blocks with inner blocks.

Additionally, the `editor` package feels like an appropriate location
for the offline detection component.

dcalhoun committed:
refactor: Fix HoC name reference

This was renamed in a commit onto which this branch was rebased.

dcalhoun committed:
refactor: Enable Fast Refresh for Image component

Place non-React component exports in a separate module to enable Fast
Refresh of the React component.

dcalhoun committed:
feat: Update the image offline message

Match the latest design comps.

dcalhoun committed:
refactor: Avoid unnecessary inconsistency in switch case logic

Replace early return with logic mirroring the rest of the switch case
statements. While I suppose it could be considered subjective, it is
odd and confusing that one of the three cases was structured
differently.

dcalhoun committed:
refactor: Enable Fast Refresh for Image component

Place non-React component exports in a separate module to enable Fast
Refresh of the React component.

dcalhoun committed:
feat: Revert logic intended to display progress bar when offline

I believe this logic erroneously prevented the spinner progress bar from
displaying when an image is successfully uploading. I.e., the logic read
"if show spinner and online, apply 'progress bar hidden' styles." I
imagine this confusion is a result of the previous logic using the `||`
pattern to conditionally apply the hidden styles, which is confusing
in itself.

This logic should likely be implemented by the way the host app
communicates the upload state. Implementing this in the JavaScript logic
is confusing or misleading, in my opinion. It adds a new, somewhat
surprising interpretation of the existing states provided by the host
app.

dcalhoun committed:
refactor: Remove sizing from offline SVG

This allows sizing the icon for different contexts.

dcalhoun committed:
feat: Update Image block media upload visuals

Update iconography and messaging when offline or or there is a upload
error.

dcalhoun committed:
feat: Add paused media upload state

Relying upon the network connection status alone results in the error
message erroneously changing whenever a network connection is
re-established. The error message, in fact, is a state communicated from
the host app's attempt to upload media. Therefore, this creates a new
pause state to represent uploads paused while offline.

derekblank committed:
Merge branch 'trunk' into rnmobile/media-upload-progress-connection-subscription

derekblank commented:
> While researching for https://github.com/wordpress-mobile/gutenberg-mobile/issues/6406, I encountered the potentially [relevant iOS code](https://github.com/wordpress-mobile/WordPress-iOS/blob/d6ec2c98cc52c591232935114cb94b9d8eccc0b7/WordPress/Classes/ViewRelated/Gutenberg/GutenbergMediaInserterHelper.swift#L244-L249). I have not yet research the Android side of the logic, but hopefully it is somewhat similar.

This looks like a good lead for iOS. I haven't precisely identified where the same logic occurs in Android, but [onMediaSaveFailed](https://github.com/wordpress-mobile/WordPress-Android/blob/33d58d267e4acf29a498d71245f76a464c717cad/libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L1516-L1520) may be a good place to start. It may also come from [onMediaUploadFailed](https://github.com/wordpress-mobile/WordPress-Android/blob/33d58d267e4acf29a498d71245f76a464c717cad/libs/editor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java#L1416-L1432) in Aztec, but this seems less likely. Will investigate further.

> From merely glancing at that logic, it would appear it already conditionally resets the progress state based on the error type. I wonder if we might be able to modify it for our needs.

Indeed. While we can modify the style code so the progress bar persists when offline, the progress value itself resets to 0, so only the progress bar track is displayed. We could possibly modify this code to hook into the network connectivity state to not mark a media upload as failed when offline. This may be the most complex part of code remaining to resolve.

derekblank commented:
> This looks like a good lead for iOS. I haven't precisely identified where the same logic occurs in Android, but [onMediaSaveFailed](https://github.com/wordpress-mobile/WordPress-Android/blob/33d58d267e4acf29a498d71245f76a464c717cad/libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L1516-L1520) may be a good place to start. It may also come from [onMediaUploadFailed](https://github.com/wordpress-mobile/WordPress-Android/blob/33d58d267e4acf29a498d71245f76a464c717cad/libs/editor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java#L1416-L1432) in Aztec, but this seems less likely. Will investigate further.

I identified further the Android code areas we'd need to update in https://github.com/WordPress/gutenberg/pull/57476.

dcalhoun committed:
feat: Scope the paused upload state to the Image block

The planned scope of the project was to improve the upload experience of
the Image block, as it is the most frequently used media block type.
Support for additional block types could be added at a later time.

dcalhoun committed:
feat: Enable paused media uploads for Media and Text block types

Because the Media and Text block type relies upon a nested Image block,
it makes sense to enable the paused media uploads for a consistent UX.

dcalhoun committed:
perf: Avoid unnecessary state updates and callback invocations

If the upload state and progress is identical, there should be no reason
to invoke callbacks, update component state, or re-render.

derekblank committed:
Add MEDIA_UPLOAD_STATE_PAUSED to Android DeferredEvent emitter

dcalhoun committed:
feat: Add Android paused media upload bridge methods

Enables the Android host app to mirror the functionality of iOS.

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into rnmobile/media-upload-progress-connection-subscription

dcalhoun committed:
docs: Add change log entry

dcalhoun committed:
feat: Add image upload failure dark mode styles

dcalhoun commented:
@derekblank I cannot request your review on this PR as you are the original author, but I believe it would be worthwhile for you to perform the final review of the code please.

derekblank commented:
@dcalhoun For some reason, it is not letting me add myself as a reviewer, even if I assign myself to the PR -- my name simply doesn't appear in the potential reviewers list. Perhaps because I opened it, and you approved already? Anyways, I tested and reviewed a final time -- LGTM. 🚀
